doc/design.md

```diff
@@ -0,0 +1,564 @@
+
+*   [**permitted-templates**] We can reject the query because it is not in the
+    list of permitted query templates; the relevant permitted template could
+    have a parameter that limits the patient, practitioner, location, etc. (see
+    below).
+*   [**query-rewrite**] We can add a constraint to impose a location based
+    restriction, e.g., if the current CHW user has access to patients in city
+    "CityHCW":
+
+    ```
+    GET [base]/Patient?address-city=CityCHW
+    ```
+
+*   [**query-rewrite**] We can add a practitioner constraint e,g.,:
+
+    ```
+    GET [base]/Patient?general-practitioner=Practitioner/HCW-ID
+    ```
+
+*   [**post-processing**] We can retrieve all results from the FHIR store, then
+    post-process them and drop everything that does not belong to this HCW.
+
+#### Sensitive data
+
+Now let's consider access to _sensitive data_. Assume that the HIV status is a
+sensitive Observation that we don't want to expose to some HCWs. What should we
+do for the following query that fetches all Observations of a patient: "`GET
+/Observation?subject=Patient/PID`"?
+
+*   [**query-rewrite**] Add the constraint to drop viral load for HIV:
+
+    "`GET /Observation?subject=Patient/PID&code:not=http://loinc.org|59419-2`"
+
+*   [**permitted-templates**] Similar to the above "all patients" example.
+*   [**post-processing**] From the returned results, remove all Observation
+    resources that have HIV viral load code.
+
+Note for the first two options we should probably use a
+[ValueSet](https://hl7.org/fhir/valueset.html) if there are a lot of Observation
+codes that we want to protect.
+
+#### Reverse chaining
+
+Finally, to see how this problem can get complicated quickly, consider this
+query:
+
+"`GET
+[base]/Patient?_id=PID&_has:Observation:patient:code-value-quantity=http://loinc.org|59419-2$gt1000||`".
+
+This query is a _[reverse chaining](https://www.hl7.org/fhir/search.html#has)_
+query which returns the queried Patients if they had a high HIV viral load.
+Post-processing cannot help here in hiding the HIV status information because
+the returned resources are Patients, not Observations. Also query-rewriting is
+difficult as we have to understand the full semantics of the query; so we cannot
+just add a simple query parameter like previous examples (this is the reason we
+have disabled "chaining" in the first release).
+
+In the main text, we refer to these examples by "all-patients",
+"sensitive-data", and "reverse-chaining" queries.
+
+## Notes
+
+[^1]: The simplified [Implicit](https://smilecdr.com/docs/smart/smart_on_fhir_authorization_flows.html#launch-flow-implicit-grant) flow could work for our use-case too but that has important security shortcomings. For example, it exposes access_token in URLs which can leak through browser history. Another more important shortcoming is that we cannot implement PKCE in the Implicit flow as the access_token is directly returned in the first request.
```

Comment:
This is actually nice, I should have used it more to make the main text more concise :-)